File: observer/exercise/BlogManagement.java
package observer.exercise;

import observer.EventType;
import observer.Offer;
import observer.Product;
import observer.Subscriber;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class BlogManagement {

    private Map<SubscriptionType, List<Subscriber>> subscribers;
    private List<Blog> blogs;
    private List<Newsletter> newsletters;

    public BlogManagement() {
        this.blogs = new ArrayList<>();
        this.newsletters = new ArrayList<>();
        initSubscribers();
    }
    
    public void writeNewBlog(Blog blog) {
        blogs.add(blog);
        notifySubscribers(SubscriptionType.NEW_BLOGS, "New blog is published...");
    }

    public void writeNewsletter(Newsletter newsletter) {
        newsletters.add(newsletter);
        notifySubscribers(SubscriptionType.NEWSLETTER, "New newsletter is published...");
    }

    private void notifySubscribers(SubscriptionType subscriptionType, String message) {
        subscribers.get(subscriptionType).forEach(subscriber -> subscriber.notify(message));
    }

    public void subscribe(SubscriptionType subscriptionType, Subscriber subscriber) {
        subscribers.get(subscriptionType).add(subscriber);
    }

    public void unsubscribe(SubscriptionType subscriptionType, Subscriber subscriber) {
        subscribers.get(subscriptionType).remove(subscriber);
    }

    private void initSubscribers() {
        subscribers.put(SubscriptionType.NEW_BLOGS, new ArrayList<>());
        subscribers.put(SubscriptionType.NEWSLETTER, new ArrayList<>());
    }
}

File: typescript/patterns/observer/exercise/blog-management.ts
import { Blog } from "./blog";
import { Newsletter } from "./newsletter";
import type { Subscriber } from "./subscriber";
import type { SubscriptionType } from "./subscription-type";

export class BlogManagement {
  private subscribers: Record<SubscriptionType, Subscriber[]>;
  private blogs: Blog[];
  private newsletters: Newsletter[];

  constructor() {
    this.subscribers = {
      NEW_BLOGS: [],
      NEWSLETTER: [],
    };
    this.blogs = [];
    this.newsletters = [];
  }

  writeNewBlog(blog: Blog) {
    this.blogs.push(blog);
    this.notifySubscribers("NEW_BLOGS", `New blog is published...`);
  }

  writeNewsletter(newsletter: Newsletter) {
    this.newsletters.push(newsletter);
    this.notifySubscribers("NEWSLETTER", `New newsletter is published...`);
  }

  private notifySubscribers(
    subscriptionType: SubscriptionType,
    message: string
  ) {
    this.subscribers[subscriptionType].forEach((subscriber) =>
      subscriber.notify(message)
    );
  }

  subscribe(subscriptionType: SubscriptionType, subscriber: Subscriber) {
    this.subscribers[subscriptionType].push(subscriber);
  }

  unsubscribe(subscriptionType: SubscriptionType, subscriber: Subscriber) {
    this.subscribers[subscriptionType] = this.subscribers[
      subscriptionType
    ].filter((s) => s !== subscriber);
  }
}
